README.md
# GUS API

**G** eoJSON   
**U** r  
**S** preadsheet   
:smile:

API for converting a Google Spreadsheet into a GeoJSON object. Inspired by the [GUS map project](https://github.com/mapsam/gus), but made to be a more generic tool to enable any type of map rendering library. Have a look at the [example spreadsheet](https://docs.google.com/spreadsheets/d/1ctA2wUBHGjrBRhQlma_x2Q5oRsC5L85XALfhkezsQMY/edit#gid=0) and an [example request](https://e53r0f186h.execute-api.us-west-2.amazonaws.com/production/1ctA2wUBHGjrBRhQlma_x2Q5oRsC5L85XALfhkezsQMY) to the API.

# Prepare your spreadsheet

The main endpoint is `GET /{spreadsheet id}` which gets your Google Spreadsheet and converts it into a GeoJSON response.

### Spreadsheet structure

The format of your spreadsheet requires the first row to be headers and two columns representing longitude and latitude. Each of these columns must be in WGS84 coordinates and can only represent points. The name of these columns MUST match the following formats:

* `longitude, latitude` or `LONGITUDE, LATITUDE`
* `long, lat` or `LONG, LAT`
* `lng, lat` or `LNG, LAT`
* `lon, lat` or `LON, LAT`
* `x, y` or `X, Y`

### Publish your spreadsheet

In order to make the spreadsheet available to the API, you must "publish to the web" by selecting `File` > `Publish to the web...` > Press `Publish`. This will make your spreadsheet publicly available. Once this is complete, grab the ID of the spreadsheet in the URL. It will look similar to this `1ctA2wUBHGjrBRhQlma_x2Q5oRsC5L85XALfhkezsQMY`. Use this ID to pass into the following endpoint

# Usage

The conversion can happen in three ways:

### API

Use the following URL to make a request (replace `{id}` with your spreadsheet ID)

```
https://e53r0f186h.execute-api.us-west-2.amazonaws.com/production/{id}
```

You'll see four possible responses:

* 200 - successful request with a GeoJSON object
* 400 - missing ID parameter
* 404 - could not find the spreadsheet, either it hasn't been published or the ID is incorrect
* 5xx - an unexpected failure with the application

### Node.js module

Install the module
```
npm install gus-api
```

Require it
```javascript
const toGeojson = require('gus-api').toGeojson;
const spreadsheetID = 'my-id';

toGeojson(id, (err, geojson) => {
  // ...
});
```

### Copy and paste

Or just copy and paste the `toGeojson` function directly from index.js!

# Develop

GUS API is deployed with AWS [API Gateway and a Lambda Proxy](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-create-api-as-simple-proxy-for-lambda.html). If you have the proper development keys, make your changes and commit, and deploy the code by running `npm run deploy` which will install node modules using the `--production` flag and zip up the directory to save on AWS s3 for the lambda to use.

## Test

```
npm test
```


CHANGELOG.md
## 0.2.0

* Enable CORS headers

## 0.1.0

* first!


test/index.test.js
const test = require('tape');
const sinon = require('sinon');
const request = require('request');
const api = require('..').handler;
const toGeojson = require('..').toGeojson;

test('[handler] returns status code 400 with no ID present', (assert) => {
  const event = {
    pathParameters: {}
  };

  api(event, {}, (err, response) => {
    assert.notOk(err);
    assert.equal(response.statusCode, 400, 'expected status code');
    assert.equal(response.body, '{\"message\":\"no spreadsheet ID provided\"}');
    assert.end();
  });
});

test('[handler] returns status code 404 when spreadsheet is not found', (assert) => {
  const event = {
    pathParameters: {
      id: 'invalid-id-123'
    }
  };

  sinon.stub(request, 'get').yields(null, { statusCode: 404 }, {});

  api(event, {}, (err, response) => {
    assert.notOk(err);
    assert.equal(request.get.firstCall.args[0], 'https://spreadsheets.google.com/feeds/cells/invalid-id-123/od6/public/basic?alt=json', 'expected url');
    assert.equal(response.statusCode, 404, 'expected status code');
    assert.equal(response.body, '{\"message\":\"unable to find spreadsheet with id: invalid-id-123\"}');
    request.get.restore();
    assert.end();
  });
});

test('[handler] returns expected geojson', (assert) => {
  const fixture = JSON.stringify(require('./fixtures/national-parks.json'));
  const event = {
    pathParameters: {
      id: 'test-id'
    }
  };

  const expected = '{"type":"FeatureCollection","features":[{"type":"Feature","geometry":{"type":"Point","coordinates":[-123.5361,47.7798]},"properties":{"longitude":"-123.5361","latitude":"47.7798","name":"Olympic National Park","description":"Rainforests, beaches, mountains, prairies"}},{"type":"Feature","geometry":{"type":"Point","coordinates":[-121.7148,46.8579]},"properties":{"longitude":"-121.7148","latitude":"46.8579","name":"Mount Rainier National Park","description":"The lonely mountain"}}]}';

  sinon.stub(request, 'get').yields(null, { statusCode: 200 }, fixture);

  api(event, {}, (err, response) => {
    assert.notOk(err);
    assert.equal(request.get.firstCall.args[0], 'https://spreadsheets.google.com/feeds/cells/test-id/od6/public/basic?alt=json', 'expected url');
    assert.equal(response.statusCode, 200, 'expected status code');
    assert.equal(response.body, expected, 'expected body');
    request.get.restore();
    assert.end();
  });
});

test('[handler] returns 400 with no lng/lat headers', (assert) => {
  const fixture = JSON.stringify(require('./fixtures/no-lng-lat-headers.json'));
  const event = {
    pathParameters: {
      id: 'test-id'
    }
  };

  sinon.stub(request, 'get').yields(null, { statusCode: 200 }, fixture);

  api(event, {}, (err, response) => {
    assert.notOk(err);
    assert.equal(response.statusCode, 400, 'expected status code');
    request.get.restore();
    assert.end();
  });
});

test('[toGeojson] fails without lng/lat headers', (assert) => {
  const fixture = JSON.stringify(require('./fixtures/no-lng-lat-headers.json'));
  const event = {
    pathParameters: {
      id: 'test-id'
    }
  };

  sinon.stub(request, 'get').yields(null, { statusCode: 200 }, fixture);

  toGeojson('fake-id', (err, gj) => {
    assert.ok(err);
    assert.equal(err.code, 'EINVALID', 'expected error code');
    assert.ok(/longitude and\/or latitude columns are missing or not properly named/.test(err.message), 'expected message');
    request.get.restore();
    assert.end();
  });
});

test('[toGeojson] inserts 0 default for missing lng/lat values in entry', (assert) => {
  const fixture = JSON.stringify(require('./fixtures/single-entry-missing-lnglat.json'));
  const event = {
    pathParameters: {
      id: 'test-id'
    }
  };

  sinon.stub(request, 'get').yields(null, { statusCode: 200 }, fixture);

  toGeojson('test-id', (err, gj) => {
    assert.notOk(err);
    assert.equal(gj.features[0].geometry.coordinates[1], 0, 'expected zero default value');
    request.get.restore();
    assert.end();
  });
});


standalone.js
function fetchJSONFile(path, callback) {
    var httpRequest = new XMLHttpRequest();
    httpRequest.onreadystatechange = function() {
        if (httpRequest.readyState === 4) {
            if (httpRequest.status === 200) {
                var data = JSON.parse(httpRequest.responseText);
                if (callback) callback(data);
            }
        }
    };
    httpRequest.open('GET', path);
    httpRequest.send();
}

function gus_api(id, callback) {
  const url = `https://spreadsheets.google.com/feeds/cells/${id}/od6/public/basic?alt=json`;

  fetchJSONFile(url, function(data) {

    let headers = {};
    let entries = {};

    data.feed.entry.forEach((e) => {
      // get the row number
      const row = parseInt(e.title['$t'].match(/\d+/g)[0]);
      const column = e.title['$t'].match(/[a-zA-Z]+/g)[0];
      const content = e.content['$t'];

      // it's a header
      if (row === 1) {
        headers[column] = content;
      } else {
        if (!entries[row]) entries[row] = {};
        entries[row][headers[column]] = content;
      }
    });

    const gj = { type: 'FeatureCollection', features: [] };
    for (let e in entries) {

      const feature = {
        type: 'Feature',
        geometry: {
          type: 'Point',
          coordinates: [0, 0]
        },
        properties: entries[e]
      };

      for (let p in entries[e]) {
        switch(p) {
          case 'longitude':
          case 'LONGITUDE':
          case 'long':
          case 'LONG':
          case 'lng':
          case 'LNG':
          case 'lon':
          case 'LON':
          case 'x':
          case 'X':
            feature.geometry.coordinates[0] = parseFloat(entries[e][p]);
          case 'latitude':
          case 'LATITUDE':
          case 'lat':
          case 'LAT':
          case 'y':
          case 'Y':
            feature.geometry.coordinates[1] = parseFloat(entries[e][p]);
        }
      }

      gj.features.push(feature);
    }

    callback(gj);
  });
};


index.js
'use strict'

const request = require('request');

const toGeojson = (id, callback) => {
  if (!id) throw new Error('No spreadsheet id provided');
  const url = `https://spreadsheets.google.com/feeds/cells/${id}/od6/public/basic?alt=json`;

  request.get(url, (err, res, body) => {
    if (err) return callback(err);
    if (res.statusCode !== 200) {
      const error = new Error(`unable to find spreadsheet with id: ${id}`);
      error.code = 'ENOENT';
      return callback(error);
    }

    body = JSON.parse(body);
    let headers = {};
    let entries = {};

    body.feed.entry.forEach((e) => {
      // get the row number
      const row = parseInt(e.title['$t'].match(/\d+/g)[0]);
      const column = e.title['$t'].match(/[a-zA-Z]+/g)[0];
      const content = e.content['$t'];

      if (row === 1) { // it's a header
        headers[column] = content;
      } else {
        if (!entries[row]) entries[row] = {};
        entries[row][headers[column]] = content;
      }
    });

    // check headers for lng/lat values
    let hasLng = false;
    let hasLat = false;
    for (let h in headers) {
      const val = headers[h];
      if (val === 'longitude' ||
          val === 'LONGITUDE' ||
          val === 'long' ||
          val === 'LONG' ||
          val === 'lng' ||
          val === 'LNG' ||
          val === 'lon' ||
          val === 'LON' ||
          val === 'x' ||
          val === 'X') hasLng = true;
      if (val === 'latitude' ||
          val === 'LATITUDE' ||
          val === 'lat' ||
          val === 'LAT' ||
          val === 'y' ||
          val === 'Y') hasLat = true;
    }

    if (!hasLng || !hasLat) {
      const error = new Error('longitude and/or latitude columns are missing or not properly named');
      error.code = 'EINVALID';
      return callback(error);
    }

    const gj = { type: 'FeatureCollection', features: [] };

    for (let e in entries) {
      const feature = {
        type: 'Feature',
        geometry: {
          type: 'Point',
          coordinates: [0, 0]
        },
        properties: entries[e]
      };

      for (let p in entries[e]) {
        switch(p) {
          case 'longitude':
          case 'LONGITUDE':
          case 'long':
          case 'LONG':
          case 'lng':
          case 'LNG':
          case 'lon':
          case 'LON':
          case 'x':
          case 'X':
            feature.geometry.coordinates[0] = parseFloat(entries[e][p] || 0);
          case 'latitude':
          case 'LATITUDE':
          case 'lat':
          case 'LAT':
          case 'y':
          case 'Y':
            feature.geometry.coordinates[1] = parseFloat(entries[e][p] || 0);
        }
      }

      gj.features.push(feature);
    }

    return callback(null, gj);
  });
};

const handler = (event, context, callback) => {
  if (!event.pathParameters.id) return callback(null, ErrorHTTP(400, 'no spreadsheet ID provided'));
  const id = event.pathParameters.id;

  toGeojson(id, (err, gj) => {
    if (err && err.code === 'ENOENT') return callback(null, ErrorHTTP(404, err.message));
    if (err && err.code === 'EINVALID') return callback(null, ErrorHTTP(400, err.message));
    if (err) return callback(null, ErrorHTTP(502, err.message));

    const response = {
      statusCode: 200,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Credentials': true,
      },
      body: JSON.stringify(gj)
    };

    return callback(null, response);
  });
};

const ErrorHTTP = (code, message) => {
  return {
    statusCode: code,
    headers: {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Credentials': true,
    },
    body: JSON.stringify({ message: message })
  };
};

module.exports = { toGeojson, handler };
